docs: Initial MVUx overview (#1251)
unoplatform/uno.extensions: doc/Overview/Reactive/overview.md

﻿---
uid: Overview.Reactive.Feed
---

# MVUx Overview

> ## What is MVU

Reviewer: This should be "What is MVUx" and should start with something along the lines of:

Model-View-Update-eXtended is and extension of the established Model-View-Update model where the flow is.......

The point is that we're not doing an in-depth on MVU, so any discussion about it should be in the context of what MVUx is

MVU stands for **M**odel-**V**iew-**U**pdate.  

It utilizes a flow where the **model** is rendered on the **view** side, and once changed by request from user, gets **update**d.

It enables and promotes leaving the actual _model_ plain and immutable,

> making the whole need of raising property change notification redundant.

Reviewer: Please leverage chatgpt or bing chat for a clearer starting point for definition of MVU:
![Uploading image.png…]()

---
uid: Overview.Reactive.Feed
---

# MVUx Overview

## What is MVU

MVU stands for **M**odel-**V**iew-**U**pdate.  

> It utilizes a flow where the *model* is rendered on the *view* side, and once changed by request from user, gets *update*d.

Author: Why *updated*  is in italic? I would keep this for feed concept like *model*  and *view* like you did (and use italic every times we are referring to those concepts)

> It enables leaving the actual model plain and immutable (ideally using records), making the whole need of raising property change notification redundant.

Author: ```suggestion
It enables leaving the actual _model_ plain and immutable records, making the whole need of raising property change notification redundant.
```
Even if here you are speaking about MVU in general, I would really make clear that records are the way to go for immutability (and feeds).

Whenever a user makes any sort of input activity that affects the displayed model, the model is then re-created, resent to the view and re-rendered,
instead of specifically notifying changes for those specific properties that were changed.

In MVU architecture, we�re rather tracking the current state of a model, and when that state changes, a new one is re-rendered.

## What is MVUx and it extends MVU

MVUx is Uno Platform's implementation of MVU. It stands for **M**odel-**V**iew-**U**pdate E**x**tended.

> MVUx extends MVU by providing powerful features that make it easier to track asynchronous requests to a remote server or other long running data-sources, and display the resulting data appropriately.

Author: I would explain here that in MVUx, the "_model_" refers only to the data, but we are not re-creating the view on each update. We are instead relying on the binding engine to reflect changes made on the _model_, which is known/expected to be more performant. It also makes MVUx UI framework independent (can be used using XAML, C# markup, but theoretically also any framework that as a binding engine)

### Metadata

> The way MVUx achieves that is by attaching metadata to the asynchronous messages sent across the pipeline.  

Author: This is only something that we offer "as a bonus" I would say, we could have implemented (a less robust) MVU without this.

Among this metadata are details about the request itself, whether it's still in progress, succeeded, or failed;
and about the returned data, letting us know if there was any data present or the response contained no data.  
The metadata also holds the previous state of the data, in case anything goes wrong and we want to retain the old state.  

> This information is then used to shape the UI accordingly, for example, if there are no data or if an error occurred.

Author: I would try to explicit a bit more all the different basic metadata we have out of the box:
* Data: <Some|None>
* Loading: <Yes|No>
* Error: <Yes|No>

We should also mention that those are extensible.

It's also a major difference compared to `IObservable`.

### Code generation

In addition, MVUx reads the model generates boilerplate source code that makes it easier to access the data and the metadata around the current state and its model, as well as asynchronous commands that can be hooked up to actions from the UI. 

### UI Controls
MVUx also provides a set of UI tools that are specially tailored to automatically read and display that metdata, providing templates for the various states, such as no data, error, progress tracking, interactions with the server to enable easy refreshing and saving of the data.

## MVUx building blocks

The following list gives an brief introduction to the main components of MVUx.

### Feed: read-only data

> MVUx uses `IFeed<T>` (where `T` is the data type) as a placeholder of an asynchronous call to a server or long-running process mainting the progress of the request and it's outcome, while preserving the previous data if anything goes wrong with the current request.

Author: ```suggestion
MVUx uses `IFeed<T>` (where `T` is the data type) as a placeholder of an asynchronous call to a server or long-running process maintaining the progress of the request and it's outcome, while preserving the previous data if anything goes wrong with the current request.
```

> An `IListFeed<T>` is an `IFeed<T>` with additional features that enable dealing with collections of `T`.

Author: ```suggestion
An `IListFeed<T>` is a feed with additional features that enable dealing with collections of `T`.
```

`IListFeed<T>` does not implement `IFeed<IList<T>>`

A feed is only used to display data received from the server and track the progress of requesting it, but it does not support changing it by the user or sending updated data back to the server, this is what State is for.

### State: read-write data

> An `IState<T>` is a superset of `IFeed<T>` which adds support for sending back updated data to the server.

Author: Currently we don't have helper to send back info to the server. I would focus instead of the fact that it allows to update the _model_
